src/vault/client/ext/aws.clj

```diff
@@ -0,0 +1,99 @@
+(ns vault.client.ext.aws
+  (:require
+    [buddy.core.codecs :refer [bytes->str]]
+    [buddy.core.codecs.base64 :as b64]
+    [cheshire.core :as json]
+    [clj-http.client :as http]
+    [clojure.java.io :as io]
+    [envoy.core :refer [defenv]]
+    [vault.client.http :as client])
+  (:import
+    (com.amazonaws
+      DefaultRequest
+      Request
+      SignableRequest)
+    (com.amazonaws.auth
+      AWS4Signer
+      AWSCredentials
+      BasicSessionCredentials
+      DefaultAWSCredentialsProviderChain)
+    com.amazonaws.http.HttpMethodName
+    java.net.URI))
+
+
+(defenv :vault-aws-iam-role
+  "The configured vault aws auth role used to perform instance authentication.")
+
+
+(def ^:private payload "Action=GetCallerIdentity&Version=2011-06-15")
+
+
+(defn sts-get-caller-identity-request
+  "Create a GetCallerIdentity request ready for signing."
+  ^SignableRequest
+  []
+  (doto (DefaultRequest. "sts")
+    (.setEndpoint (URI. "https://sts.amazonaws.com"))
+    (.setResourcePath "/")
+    (.setHttpMethod HttpMethodName/POST)
+    (.addHeader "Content-Type" "application/x-www-form-urlencoded; charset=utf-8")
+    (.addHeader "Content-Length" "43")
+    (.setContent (-> payload (.getBytes "UTF-8") (io/input-stream)))))
+
+
+(defn credentials
+  "AWSCredentials to authenticate with STS. Will fall back to default chain including instance profile."
+  ^AWSCredentials
+  ([id secret token]
+   (BasicSessionCredentials. id secret token))
```

Comment:
Seems like it would be good to also accept a 2-arg version for non-session credentials.